### PR TITLE
Removes deployment of owner DNSRecord

### PR DESCRIPTION
**Pull request by plkokanov**
**How to categorize this PR?**

/area control-plane-migration
/kind cleanup

**What this PR does / why we need it**:
Removes the deployment of owner `DNSRecord` during `Shoot` reconciliation, deletion and migration as they are no longer needed due to the removal of owner checks as part of #6302.
The deletion of the owner `DNSRecord` is kept so that they can be properly cleaned up if they were created by a previous version of gardener. 

The lack of owner `DNSRecord`s is not a problem If any extension controllers are not yet revendored to the `gardener` release that includes the owner checks removal from the `extensions` library. The owner checks are simply not started of the owner `DNSRecord` is not found.

**Which issue(s) this PR fixes**:
Part of #6302

**Special notes for your reviewer**:
Depends on https://github.com/gardener/gardener/pull/6988 and https://github.com/gardener/gardener/pull/6973.

**Release note**:

```other operator
The special owner `DNSRecord` resources are no longer deployed.
```

### CODE REVIEW

**Comment by rfranzke** [comment]
/assign

@plkokanov, one question:

> The lack of owner DNSRecords is not a problem If any extension controllers are not yet revendored to the gardener release that includes the owner checks removal from the extensions library. The owner checks are simply not started of the owner DNSRecord is not found.

What if extension controllers are still running when a new gardenlet is deployed which starts to delete the `DNSRecord`s?

**Comment by plkokanov** [comment]
> What if extension controllers are still running when a new gardenlet is deployed which starts to delete the DNSRecords?

I didn't think of that case so I will have to check. However, I think that this corner case was also present previously if you were to disable the owner checks in the seed settings and reconcile a `Shoot` cluster assigned to that seed as the reconciliation would remove the owner `DNSRecords`

/hold

**Comment by plkokanov** [comment]
Once a reconciliation for an extension resource in a namespace starts, a `watchdog` is created that tries to resolve the owner `DNSRecord`. This watchdog is then saved in a map and a timer is registered that will delete the `watchdog` after 10 minutes. 
https://github.com/gardener/gardener/blob/d9376c117efb8f31334131cf1d995d99fc2f51d4/extensions/pkg/controller/common/watchdogmanager.go#L152-L170

If the resolution of the owner `DNSrecord` succeeds a context for the reconciliation is created, and the removal timer is canceled.
https://github.com/gardener/gardener/blob/d9376c117efb8f31334131cf1d995d99fc2f51d4/extensions/pkg/controller/common/watchdogmanager.go#L125-L138

Eventually when the reconciliation finishes or errors, the context for it is removed and the removal timer is restarted if it was the last context tracked by the watchdog
https://github.com/gardener/gardener/blob/d9376c117efb8f31334131cf1d995d99fc2f51d4/extensions/pkg/controller/common/watchdogmanager.go#L139-L144

One watchdog is created per control plane namespace and extension controller. (e.g one watchdog is used for `shoot--foo--bar` namespace and all extension resources reconciled by the provider-aws extension)

So, as long as there was no reconciliation for the last 10 minutes, the removal of the DNS records would be ok. However, if an extension reconciliation is stuck in a crashloop backoff, the context will be constantly readded and the removal timer cancelled until the backoff becomes greater than 10 minutes (if that's possible).

File: pkg/gardenlet/controller/shoot/shoot_control_reconcile.go [hunk 1]
```diff
@@ -163,6 +163,9 @@ func (r *shootReconciler) runReconcileShootFlow(ctx context.Context, o *operatio
 			Fn:           flow.TaskFn(botanist.DeployReferencedResources).RetryUntilTimeout(defaultInterval, defaultTimeout),
 			Dependencies: flow.NewTaskIDs(deployNamespace),
 		})
+		// TODO(plkokanov): This step can be removed in a future version, when all owner
+		// DNSRecords have been cleaned up from existing Shoots. Currently this step deletes
+		// the owner DNSRecord.
 		deployOwnerDomainDNSRecord = g.Add(flow.Task{
 			Name:         "Deploying owner domain DNS record",
```
Comment (timebertt) [inline pkg/gardenlet/controller/shoot/shoot_control_reconcile.go]: Please rename the task according to what it actually does

File: pkg/operation/botanist/dnsresources.go [hunk 2]
```diff
@@ -16,23 +16,21 @@ package botanist
...
-	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
 )
 
 // DeployOwnerDNSResources deploys or deletes the owner DNSRecord resource depending on whether
 // the 'ownerChecks' setting is enabled.
 // * If the ownerChecks is enabled, the DNSRecord resource is deployed (or restored).
 // * Otherwise, it is deleted.
+// TODO(plkokanov): This can be removed in a future version, when all owner
+// DNSRecords have been cleaned up from existing Shoots.
 func (b *Botanist) DeployOwnerDNSResources(ctx context.Context) error {
```
Comment (timebertt) [inline pkg/operation/botanist/dnsresources.go]: Please remove this function along with `MigrateOwnerDNSResources` and simply call `DestroyOwnerDNSRecord` directly.

**Review by timebertt — COMMENTED** [review]
I think we should require new extension versions to be rolled out that have vendored https://github.com/gardener/gardener/pull/6973.

/assign

**Comment by timebertt** [comment]
Additionally, all shoots need to be reconciled with https://github.com/gardener/gardener/pull/6988 before removing the owner DNSRecords

**Comment by MartinWeindel** [comment]
Please note: The owner `DNSRecord` is also used by the shoot-dns-service to activate its `DNSOwner`. See https://github.com/gardener/gardener-extension-shoot-dns-service/pull/90
At least, the activation of the `DNSOwner` by the owner DNSRecord must be disabled if it is removed. 
As I was not directly involved in the original design discussions, I'm not sure about the consequences on the bad case migration path.
It would be a good idea to get feedback from @mandelsoft here.

**Comment by plkokanov** [comment]
@MartinWeindel @mandelsoft as part of the discussion on #6302 we decided that bad case control plane migration is largely a corner case and it will bring very little benefit when the gardener HA topic is finished. We will still support forced migration in very rare cases (e.g. when a seed cluster is completely cut off from the internet) until HA is available, but the control plane in the source seed will be turned off via a different mechanism, e.g. by rotating the user's credentials.

**Comment by MartinWeindel** [comment]
@plkokanov Ok, thanks for the explanation. We will drop the DNS activation of the `DNSOnwer` in the shoot-dns-service as soon as possible to remove the dependency on the owner `DNSRecord`.

**Comment by plkokanov** [comment]
Some extensions have still not vendored gardener `v1.60` so this PR can not be merged yet.
I will close it for now so that it does not clutter up the other PRs

/close

**Pull request closed without merging**